### PR TITLE
RSS Виртуальная клавиатура

### PR DESCRIPTION
**Pull request by Lex-tp**
# RSS Виртуальная клавиатура #

### Скриншот
 Иллюстрация к проекту
## Минимальный набор:

- [x] Реализована генерация DOM-элементов и body в index.html пустой изначально (может находится только тег script): +20
- [x] Нажатие на кнопку на физической клавиатуре подсвечивает кнопку на виртуальной(проверять следуюет нажатие цифр, букв, знаков припинания, backspace, del(если она присутствует), enter, shift, alt, ctrl, tab, caps lock, space, стрелки вниз-вверх-влево-вправо): +10

## Cтандартный набор:

- [x] Есть переключение между русским и английским языком(сочетание клавиш для переключения языка должно быть указано на странице с клавиатурой), а так же сохранение выбранного языка: +15
- [x] Клики мышкой по кнопкам на виртуальной клавиатуре или нажатие на кнопки физической клавиатуры, выводят символы в инпут(textarea): +15

## Дополнительный набор:

- [x] Реализована анимация нажатия на кнопку: +15

## Технические требования:

- [x] Использование в коде фишек стандарта ES6 и выше (classes, деструктуризация и т.д.): +15
- [x] Использование eslint: +10
- [x] Требования к репозиторию, коммитам и PR выполнены: +10

## Штрафы:

- [x] Ошибки в консоли, связанные с исполняемым кодом (ошибки типа favicon.ico: Failed to load resource: the server responded with a status of 404 не учитываются) или предупреждения eslint-config-airbnb-base: -15

### CODE REVIEW

File: scripts/main.js [hunk 1]
```diff
@@ -0,0 +1,558 @@
...
+    { key: "Ъ", code: 221 }, { key: "/", code: 220 }, { key: "Del", code: 46 }],
+    [{ key: "Caps Lock", code: 20 }, { key: "Ф", code: 65 }, { key: "Ы", code: 83 }, { key: "В", code: 68 }, { key: "А", code: 70 }, { key: "П", code: 71 },
+    { key: "Р", code: 72 }, { key: "О", code: 74 }, { key: "Л", code: 75 }, { key: "Д", code: 76 }, { key: "Ж", code: 186 }, { key: "Э", code: 222 },
+    { key: "Enter", code: 13 }],
+    [{ key: "Shift", code: "ShiftLeft" }, { key: "Я", code: 90 }, { key: "Ч", code: 88 }, { key: "С", code: 67 }, { key: "М", code: 86 }, { key: "И", code: 66 },
+    { key: "Т", code: 78 }, { key: "Ь", code: 77 }, { key: "Б", code: 188 }, { key: "Ю", code: 190 }, { key: ",", code: 191 }, { key: "🡅", code: 38 },
+    { key: "Shift", code: "ShiftRight" }],
+    [{ key: "Ctrl", code: "ControlLeft" }, { key: "Alt", code: "AltLeft" }, { key: "Space", code: "Space" }, { key: "Alt", code: "AltRight" }, { key: "Ctrl", code: "ControlRight" },
+    { key: "🡄", code: 37 }, { key: "🡇", code: 40 }, { key: "🡆", code: 39 }, { key: "Win", code: 92 }]
+];
```
Comment (iZamza) [inline scripts/main.js:33]: why not toUpperCase for russian board? You change just key.value, codes are same

File: scripts/main.js [hunk 2]
```diff
@@ -0,0 +1,558 @@
...
+    { key: "}", code: 221 }, { key: "|", code: 220 }, { key: "Del", code: 46 }],
+    [{ key: "Caps Lock", code: 20 }, { key: "A", code: 65 }, { key: "S", code: 83 }, { key: "D", code: 68 }, { key: "F", code: 70 }, { key: "G", code: 71 },
+    { key: "H", code: 72 }, { key: "J", code: 74 }, { key: "K", code: 75 }, { key: "L", code: 76 }, { key: ":", code: 186 }, { key: "\"", code: 222 },
+    { key: "Enter", code: 13 }],
+    [{ key: "Shift", code: "ShiftLeft" }, { key: "Z", code: 90 }, { key: "X", code: 88 }, { key: "C", code: 67 }, { key: "V", code: 86 }, { key: "B", code: 66 },
+    { key: "N", code: 78 }, { key: "M", code: 77 }, { key: "<", code: 188 }, { key: ">", code: 190 }, { key: "?", code: 191 }, { key: "🡅", code: 38 },
+    { key: "Shift", code: "ShiftRight" }],
+    [{ key: "Ctrl", code: "ControlLeft" }, { key: "Alt", code: "AltLeft" }, { key: "Space", code: "Space" }, { key: "Alt", code: "AltRight" }, { key: "Ctrl", code: "ControlRight" },
+    { key: "🡄", code: 37 }, { key: "🡇", code: 40 }, { key: "🡆", code: 39 }, { key: "Win", code: 92 }]
+];
```
Comment (iZamza) [inline scripts/main.js:67]: same question

File: scripts/main.js [hunk 3]
```diff
@@ -0,0 +1,558 @@
...
+let infoText = document.createElement("span");
+let infoOS = document.createElement("span");
+let infoTextPlus = document.createElement("span");
+let infoShiftBtn = document.createElement("div");
+let infoShiftText = document.createElement("p");
+let infoCtrlBtn = document.createElement("div");
+let infoCtrlText = document.createElement("p");
+let inputText = document.createElement("div");
+let textArea = document.createElement("textarea");
+let buttonKey, key, lineBoard;
```
Comment (iZamza) [inline scripts/main.js:86]: why this place not head of document? Please, read about hoisting

File: scripts/main.js [hunk 4]
```diff
@@ -0,0 +1,558 @@
...
+    infoBox.appendChild(infoCtrlBtn);
+    infoBox.appendChild(infoOS);
+    inputText.appendChild(textArea);
+    container.appendChild(inputText);
+    container.appendChild(keyBoard);
+    container.appendChild(infoBox);
+    document.body.append(container);
+}
+
+document.body.onload=()=> {
```
Comment (iZamza) [inline scripts/main.js:120]: arrow function ES6 feature, why you not added +15 points for this in self estimation?

File: scripts/main.js [hunk 5]
```diff
@@ -0,0 +1,558 @@
...
+    if(localStorage.getItem("ENG")=="true"){
+        isEng=false;
+        changeLang();
+    }
+    else {
+        isEng=true;
+        changeLang();
+    }
+};
```
Comment (iZamza) [inline scripts/main.js:129]: better write just
`isEng = !!localStorage.getItem("ENG");
changeLang();`

File: scripts/main.js [hunk 6]
```diff
@@ -0,0 +1,558 @@
...
+    infoBox.appendChild(infoOS);
+    inputText.appendChild(textArea);
+    container.appendChild(inputText);
+    container.appendChild(keyBoard);
+    container.appendChild(infoBox);
+    document.body.append(container);
+}
+
+document.body.onload=()=> {
+    if(localStorage.getItem("ENG")=="true"){
```
Comment (iZamza) [inline scripts/main.js:121]: try avoid abstract comparison. 
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Comparison_Operators

File: scripts/main.js [hunk 7]
```diff
@@ -0,0 +1,558 @@
...
+        isEng=false;
+        changeLang();
+    }
+    else {
+        isEng=true;
+        changeLang();
+    }
+};
+
+let createButtons = function (button, keyButton, keyCode, key) {
```
Comment (iZamza) [inline scripts/main.js:131]: better use function declaration (read about hoistiong)

File: scripts/main.js [hunk 8]
```diff
@@ -0,0 +1,558 @@
...
+    button = document.createElement("div");
```
Comment (iZamza) [inline scripts/main.js:132]: not really understand why you pass something like function arguments, and after that you change it's value to `document.createElement('div'); ` I think you can use just `let button = document.createElement('div');` witjout catching something outside function (read about closure and functional, bracket scopes)

File: scripts/main.js [hunk 9]
```diff
@@ -0,0 +1,558 @@
...
+    keyBoard.appendChild(lineBoard);
+    for (let j = 0; j < RU_KEYS[i].length; j++) {
+        buttonKey = createButtons(buttonKey, key, RU_KEYS[i][j].code, RU_KEYS[i][j].key);
+        lineBoard.appendChild(buttonKey);
+    }
+}
+
+function changeCaps() {
+    if (!isCaps) {
+        if (isEng == false) {
```
Comment (iZamza) [inline scripts/main.js:172]: try avoid abstract comparison.

File: scripts/main.js [hunk 10]
```diff
@@ -0,0 +1,558 @@
...
+            keyBoard.querySelectorAll(".button").forEach(function (e) {
+                for (let j = 0; j < RU_KEYS_CAPS.length; j++) {
+                    if (RU_KEYS_CAPS[j] == e.getAttribute("data")) {
+                        e.querySelector("p").innerText = e.querySelector("p").innerText.toUpperCase();
+                    }
+                }
+            });
+        }
+        if (isEng == true) {
```
Comment (iZamza) [inline scripts/main.js:181]: try avoid abstract comparison.

File: scripts/main.js [hunk 11]
```diff
@@ -0,0 +1,558 @@
...
+        lineBoard.appendChild(buttonKey);
+    }
+}
+
+function changeCaps() {
+    if (!isCaps) {
+        if (isEng == false) {
+            keyBoard.querySelectorAll(".button").forEach(function (e) {
+                for (let j = 0; j < RU_KEYS_CAPS.length; j++) {
+                    if (RU_KEYS_CAPS[j] == e.getAttribute("data")) {
```
Comment (iZamza) [inline scripts/main.js:175]: try avoid abstract comparison.

File: scripts/main.js [hunk 12]
```diff
@@ -0,0 +1,558 @@
...
+                        e.querySelector("p").innerText = e.querySelector("p").innerText.toUpperCase();
```
Comment (iZamza) [inline scripts/main.js:176]: if you have big constructions store they into variables.
like `let elementParagraphText = e.querySelector("p").innerText` and reuse it

File: scripts/main.js [hunk 13]
```diff
@@ -0,0 +1,558 @@
...
+    for (let j = 0; j < RU_KEYS[i].length; j++) {
+        buttonKey = createButtons(buttonKey, key, RU_KEYS[i][j].code, RU_KEYS[i][j].key);
+        lineBoard.appendChild(buttonKey);
+    }
+}
+
+function changeCaps() {
+    if (!isCaps) {
+        if (isEng == false) {
+            keyBoard.querySelectorAll(".button").forEach(function (e) {
```
Comment (iZamza) [inline scripts/main.js:173]: `e `usually use like event, not element

File: scripts/main.js [hunk 14]
```diff
@@ -0,0 +1,558 @@
...
+        if (!textArea.setRangeText) { return; }
+        if (start >= end) {
+            if (start <= 0 || !textArea.setSelectionRange) { return; }
+            textArea.setSelectionRange(start - 1, start);
+        }
+        textArea.setRangeText("");
+        textArea.focus();
+    }
+    if (event.target.getAttribute("data") == 46 || selectItem.parentNode.getAttribute("data") == 46) {
+        let start = textArea.selectionStart,
```
Comment (iZamza) [inline scripts/main.js:245]: why not switch case?)

File: scripts/main.js [hunk 15]
```diff
@@ -0,0 +1,558 @@
...
+    && (event.target.getAttribute("data") != 9 && selectItem.parentNode.getAttribute("data") != 9) && (event.target.getAttribute("data") != "Space" && selectItem.parentNode.getAttribute("data") != "Space") 
+    && (event.target.getAttribute("data") != 46 && selectItem.parentNode.getAttribute("data") != 46) && (event.target.getAttribute("data") != 20 && selectItem.parentNode.getAttribute("data") != 20)
+    && (event.target.getAttribute("data") != "ControlLeft" && selectItem.parentNode.getAttribute("data") != "ControlLeft") && (event.target.getAttribute("data") != "ControlRight" && selectItem.parentNode.getAttribute("data") != "ControlRight")
+    && (event.target.getAttribute("data") != "AltLeft" && selectItem.parentNode.getAttribute("data") != "AltLeft") && (event.target.getAttribute("data") != "AltRight" && selectItem.parentNode.getAttribute("data") != "AltRight")
+    && (event.target.getAttribute("data") != 92  && selectItem.parentNode.getAttribute("data") != 92)) {
+        event.target.tagName == "DIV" ? setSelectionRange(selectItem.textContent) : setSelectionRange(selectItem.parentNode.textContent);
+    }
```
Comment (iZamza) [inline scripts/main.js:270]: not readable

File: scripts/main.js [hunk 16]
```diff
@@ -0,0 +1,558 @@
...
+                    if (EN_KEYS_UP[0][j].code == e.getAttribute("data")) {
+                        e.querySelector("p").innerText = EN_KEYS_UP[0][j].key;
+                    }
+                }
+            });
+        }
+    }
+});
+
+document.addEventListener("mouseup", function (event) {
```
Comment (iZamza) [inline scripts/main.js:314]: better add event listeners to some element not for document object

File: scripts/main.js [hunk 17]
```diff
@@ -0,0 +1,558 @@
...
+    textArea.focus();
+    textArea.setRangeText(text, textArea.selectionStart, textArea.selectionEnd, "end");
+
+}
+
+document.body.addEventListener("keyup", function (event) {
+    keyBoard.querySelectorAll(".button").forEach(function (e) {
+        e.classList.remove("active");
+    });
+    document.querySelector(`.button[data="${event.code}"]`) === null ? null : document.querySelector(`.button[data="${event.code}"]`).classList.remove("active-shift");
```
Comment (iZamza) [inline scripts/main.js:512]: if null => null ?)
`if (document.querySelector(`.button[data="${event.code}"]`)) {document.querySelector(`.button[data="${event.code}"]`).classList.remove("active-shift");}`

File: styles/css/main.css [hunk 18]
```diff
@@ -1,3 +1,164 @@
...
+  line-height: 10px;
+  border: 1px solid white;
+  margin-right: 5px;
+  font-family: 'Roboto', sans-serif;
+  color: white;
+}
+.button-big-space:hover {
+  background-color: #188FFF;
+  border: 1px solid white;
+  color: white;
```
Comment (iZamza) [inline styles/css/main.css:68]: classes looks same, why you don't uses multyclasses in DOM for it. You will have one general class (button). and for button-space add properties that will be different

File: styles/scss/main.scss [hunk 19]
```diff
@@ -0,0 +1,148 @@
...
+    font-family: 'Roboto', sans-serif;
+    color: white;
+    &:hover {
+        background-color: $hoverColor;
+        border: 1px solid white; 
+        color: white;
+        cursor: pointer;
+        translate: 1s liner;
+    }
+}
```
Comment (iZamza) [inline styles/scss/main.scss:71]: into
```
.button-big{
          &-space {...}
}
```

**Review by iZamza — COMMENTED** [review]
Dear Denis, you make good job, but also you have opportunity to grow up! Good lack!)

File: scripts/main.js (hunk 1 above)
Comment (Lex-tp) [reply scripts/main.js:33]: > why not toUpperCase for russian board? You change just key.value, codes are same

Thanks for the comment. Fixed and combined the array into one.

File: scripts/main.js (hunk 2 above)
Comment (Lex-tp) [reply scripts/main.js:67]: > same question

Thanks for the comment. Also fixed and merged the array into one.

File: scripts/main.js (hunk 6 above)
Comment (Lex-tp) [reply scripts/main.js:121]: > try avoid abstract comparison.
> https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Comparison_Operators

Thanks for your comment. Corrected and remembered.

File: scripts/main.js (hunk 8 above)
Comment (Lex-tp) [reply scripts/main.js:132]: > not really understand why you pass something like function arguments, and after that you change it's value to `document.createElement('div'); ` I think you can use just `let button = document.createElement('div');` witjout catching something outside function (read about closure and functional, bracket scopes)

Thank you for your comment. I agree this is very stupid.

File: scripts/main.js (hunk 13 above)
Comment (Lex-tp) [reply scripts/main.js:173]: > `e `usually use like event, not element

Thank you for your comment. I will remember this.

File: scripts/main.js (hunk 14 above)
Comment (Lex-tp) [reply scripts/main.js:245]: > why not switch case?)

Thank you for your comment. Switch cannot have two parameters

File: scripts/main.js (hunk 17 above)
Comment (Lex-tp) [reply scripts/main.js:512]: > if null => null ?)
> `if (document.querySelector(`.button[data="${event.code}"]`)) {document.querySelector(`.button[data="${event.code}"]`).classList.remove("active-shift");}`

Thank you for your comment. I agree that this is also very stupid.

File: styles/scss/main.scss (hunk 19 above)
Comment (Lex-tp) [reply styles/scss/main.scss:71]: > into
> 
> ```
> .button-big{
>           &-space {...}
> }
> ```

Thank you for your comment. I tried to fix

File: styles/css/main.css (hunk 18 above)
Comment (Lex-tp) [reply styles/css/main.css:68]: > classes looks same, why you don't uses multyclasses in DOM for it. You will have one general class (button). and for button-space add properties that will be different

Thank you for your comment. I tried to fix

**Pull request merged**